code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm

```diff
@@ -0,0 +1,529 @@
+ * goes into warning mode if gas parameters are beyond the tlv warning bounds, goes into hazard mode if gas parameters are beyond tlv hazard bounds
+ *
+ */
+/obj/machinery/airalarm/proc/check_air_dangerlevel(turf/location, datum/gas_mixture/environment, exposed_temperature)
+	SIGNAL_HANDLER
+	if((machine_stat & (NOPOWER|BROKEN)) || shorted)
+		return
+
+	var/datum/tlv/current_tlv
+	//cache for sanic speed (lists are references anyways)
+	var/list/cached_tlv = tlv_collection
+
+	var/list/env_gases = environment.gases
+	var/partial_pressure = R_IDEAL_GAS_EQUATION * exposed_temperature / environment.volume
+
+	current_tlv = cached_tlv["pressure"]
+	var/environment_pressure = environment.return_pressure()
+	var/pressure_dangerlevel = current_tlv.check_value(environment_pressure)
+
+	current_tlv = cached_tlv["temperature"]
+	var/temperature_dangerlevel = current_tlv.check_value(exposed_temperature)
+
+	var/gas_dangerlevel = 0
+	for(var/gas_id in env_gases)
+		if(!(gas_id in cached_tlv)) // We're not interested in this gas, it seems.
+			continue
+		current_tlv = cached_tlv[gas_id]
+		gas_dangerlevel = max(gas_dangerlevel, current_tlv.check_value(env_gases[gas_id][MOLES] * partial_pressure))
+
+	environment.garbage_collect()
+
+	var/old_danger_level = danger_level
+	danger_level = max(pressure_dangerlevel, temperature_dangerlevel, gas_dangerlevel)
+
+	if(old_danger_level != danger_level)
+		INVOKE_ASYNC(src, PROC_REF(apply_danger_level))
+	if(istype(selected_mode, /datum/air_alarm_mode/cycle) && environment_pressure < ONE_ATMOSPHERE * 0.05)
+		var/datum/air_alarm_mode/cycle/typed_mode = selected_mode
+		typed_mode.replace(my_area)
+
+/obj/machinery/airalarm/proc/apply_danger_level()
+
+	var/new_area_danger_level = 0
+	for(var/obj/machinery/airalarm/AA in my_area)
+		if (!(AA.machine_stat & (NOPOWER|BROKEN)) && !AA.shorted)
+			new_area_danger_level = clamp(max(new_area_danger_level, AA.danger_level), 0, 1)
+
+	var/did_anything_happen
+	if(new_area_danger_level)
+		did_anything_happen = alarm_manager.send_alarm(ALARM_ATMOS)
+	else
+		did_anything_happen = alarm_manager.clear_alarm(ALARM_ATMOS)
+	if(did_anything_happen) //if something actually changed
+		danger_level = new_area_danger_level
+
+	update_appearance()
+
+/obj/machinery/airalarm/proc/select_mode(atom/source, datum/air_alarm_mode/mode_path)
+	var/datum/air_alarm_mode/new_mode = GLOB.air_alarm_modes[mode_path]
+	if(!new_mode)
+		return
+	if(new_mode.emag && !(obj_flags & EMAGGED))
+		return
+	selected_mode = new_mode
+	selected_mode.apply(my_area)
+	SEND_SIGNAL(src, COMSIG_AIRALARM_UPDATE_MODE, source)
+
+MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/airalarm, 24)
+
+/proc/extract_id_tags(list/objects)
+	var/list/tags = list()
+
+	for (var/obj/object as anything in objects)
+		tags += object.id_tag
+
+	return tags
+
+/proc/find_by_id_tag(list/objects, id_tag)
+	for (var/obj/object as anything in objects)
+		if (object.id_tag == id_tag)
+			return object
+
+	return null
```

Comment:
should these be here? __HELPERS file maybe?

Reply:
We'll claim it as one of atmos'